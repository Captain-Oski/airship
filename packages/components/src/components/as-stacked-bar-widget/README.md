
TODO: Description

### Props


#### **heading**: string
Heading text of the widget

```code
lang: html
---
<as-stacked-bar-widget heading="Heading"></as-stacked-bar-widget>
```
```code
lang: javascript
---
stackedBar.heading = 'Heading';
```

#### **description**: string
Description text of the widget

```code
lang: html
---
<as-stacked-bar-widget description="Description"></as-stacked-bar-widget>
```
```code
lang: javascript
---
stackedBar.description = 'Description';
```
