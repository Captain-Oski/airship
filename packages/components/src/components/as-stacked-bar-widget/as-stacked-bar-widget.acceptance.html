<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <base href="/">
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
  <title>Airship</title>
  <link rel="stylesheet" href="/dist/airship.css">
  <script src="/build/airship.js"></script>
  <style>
    body {
      background: #CCC;
    }

    as-stacked-bar-widget {
      display: inline-block;
      background: white;
    }

    #widget-0,
    #widget-1,
    #widget-2 {
      /* width: 350px; */
    }

    .sidebar-xl {
      width: 640px;
    }
  </style>
</head>

<body class="as-app">

  <header class="as-toolbar"></header>
  <as-responsive-content>
    <aside class="as-sidebar as-sidebar--right sidebar-xl">
      <as-stacked-bar-widget id="widget-0" show-legend="true" heading="Heading" description="Description"></as-stacked-bar-widget>
      <as-histogram-widget
        heading="Title"
        description="Description"
        show-header
        show-clear>
      </as-histogram-widget>
    </aside>

    <main class="as-main">
      <div class="as-map-area">
        <div id="map"></div>
        <div class="as-map-panels">
          <div class="as-panel as-panel--top as-panel--right">
            <div class="as-panel__element"></div>
          </div>
        </div>
      </div>
      <footer class="as-map-footer as-bg--complementary">
        <as-stacked-bar-widget id="widget-1" show-legend="true" heading="Heading" description="Description"></as-stacked-bar-widget>
        <as-histogram-widget
          id="histo2"
          heading="Title"
          description="Description"
          show-header
          show-clear>
        </as-histogram-widget>

      </footer>
    </main>
  </as-responsive-content>

  <script>
    const widget0 = document.querySelector('#widget-0');
    var data = [{
        category: 'Category A',
        values: {
          v1: 100,
          v2: 100,
        }
      },
      {
        category: 'Category B',
        values: {
          v1: 100,
          v2: 200,
          v3: 300,
        },
      },
      {
        category: 'Category B',
        values: {
          v1: 100,
          v2: 200,
          v3: 300,
          v3: 400,
        }
      }
    ];
    widget0.data = data;

    var metadata = {
      v1: {
        label: 'v1-label',
        color: '#FABADA'
      },
      v2: {
        label: 'v2-label',
        color: '#C00FEE'
      },
      v3: {
        color: '#ff0088',
      },
    };
    widget0.metadata = metadata;

    const widget1 = document.querySelector('#widget-1');
    widget1.data = [{
        category: 'Category A',
        values: {
          v1: 1000,
          v2: 1000,
        }
      },
      {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
        },
      },
      {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        }
      },
      {
        category: 'Category A',
        values: {
          v1: 1000,
          v2: 1000,
        }
      },
      {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
        },
      },
      {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      }, {
        category: 'Category B',
        values: {
          v1: 1000,
          v2: 2000,
          v3: 3000,
          v3: 4000,
        },
      },
    ];

  var histogramWidget = document.querySelector('as-histogram-widget');
  var histogramWidget2 = document.getElementById('histo2');
  var histoData = [
    { start: 0, end: 10, value: 5 },
    { start: 10, end: 20, value: 10 },
    { start: 20, end: 30, value: 15 },
    { start: 30, end: 40, value: 20 },
    { start: 40, end: 50, value: 30 },
  ];
  histogramWidget.data = histoData;
  histogramWidget2.data = histoData;

  var data1 = [
    {
      category: 'Category A',
      values: {
        v1: 100,
        v2: 100,
      }
    },
    {
      category: 'Category B',
      values: {
        v1: 100,
        v2: 200,
        v3: 300,
        v3: 400,
      }
    }
  ];

  var data2 = [
    {
      category: 'Category A',
      values: {
        v1: -30,
        v2: 40,
      }
    },
    {
      category: 'Category B',
      values: {
        v1: 10,
        v2: 20,
        v3: 30,
      },
    },
    {
      category: 'Category B',
      values: {
        v1: 50,
        v2: 200,
        v3: 300,
        v3: 40,
      }
    }
  ];

    var metadata1 = {
      v1: {
        label: 'tucutu',
        color: '#FF0000'
      },
      v2: {
        label: 'tacata',
        color: '#00FF00'
      },
      v3: {
        color: '#0000FF',
      },
    };

  </script>
</body>

</html>
