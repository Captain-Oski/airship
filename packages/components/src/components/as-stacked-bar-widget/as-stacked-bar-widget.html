<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <base href="/">
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
  <title>Airship</title>
  <link rel="stylesheet" href="/dist/airship.css">
  <script src="/build/airship.js"></script>
  <style>
    body {
      background: #f2f2f2;
    }

  </style>
</head>

<body>
  <as-stacked-bar-widget show-legend="true" heading="Heading" description="Description"></as-stacked-bar-widget>

  <script>
    const widget = document.querySelector('as-stacked-bar-widget');
    widget.data = [{
        category: 'A New Hope',
        values: {
          investment: -900,
          revenue: 5000
        }
      },
    ];

    console.log(widget.data);

  </script>

</body>

</html>
