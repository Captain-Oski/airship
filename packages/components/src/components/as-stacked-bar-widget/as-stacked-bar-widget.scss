@import '../../../../styles/src/core/_dev';
@import '../../../../styles/src/tooltip/tooltip';

as-stacked-bar-widget {
  display: block;
  width: 100%;
  padding: 16px;
  background: var(--histogram-widget--background-color, $color-ui-01);

  .as-stacked-bar-widget__header {
    margin: 0;
    font: 16px/24px var(--font-family-base, Roboto, sans-serif);
  }

  .as-stacked-bar-widget__description {
    margin-bottom: 16px;
    color: var(--histogram-widget--description-color, #747474);
  }

  svg {
    width: 100%;
    padding: 16px;
    overflow: visible;

    .plot {
      transform: translate(25px, 0) scaleY(0.98);
      transform-origin: center;

      rect {
        cursor: pointer;
      }
    }
  }

  as-legend {
    margin-left: 16px;
  }

  [role="tooltip"] {
    display: none;
    position: fixed;
    transform: translate(-50%, -120%);
    pointer-events: none;
  }
}
