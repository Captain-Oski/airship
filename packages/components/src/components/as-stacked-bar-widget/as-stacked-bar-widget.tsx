import { Component, Prop } from '@stencil/core';
import { select, Selection } from 'd3-selection';
import dataProcessor from './data-processor';


/**
 * Stacked bar Widget
 *
 * @export
 * @class StackedBarWidget
 */
@Component({
  shadow: false,
  styleUrl: './as-stacked-bar-widget.scss',
  tag: 'as-stacked-bar-widget',
})
export class StackedBarWidget {

  /**
   * Header of the widget to be displayed
   *
   * @type {string}
   * @memberof StackedBarWidget
   */
  @Prop() public heading: string;

  /**
   * Description of the widget to be displayed
   *
   * @type {string}
   * @memberof StackedBarWidget
   */
  @Prop() public description: string;

  /**
   * Boolean flag to control legend visibility.
   * Defaults: true
   *
   * @type {boolean}
   * @memberof StackedBarWidget
   */
  @Prop() public showLegend: boolean = true;

  @Prop() public data: StackedbarData[] = [];

  private container: Selection<HTMLElement, {}, null, undefined>;

  public render() {
    const [from, to] = dataProcessor.getDomain(this.data);
    return (
      <div class="as-widget">
        {this._renderHeader()}
        <svg ref={(ref: HTMLElement) => this.container = select(ref)}></svg>
        <as-y-axis from={from} to={to}></as-y-axis>
        {this._renderLegend()}
      </div>
    );
  }

  public componentDidLoad() {
    // this._renderGraph();
  }

  private _renderHeader() {
    return [
      <h2 class='as-stacked-bar-widget__header'>{this.heading}</h2>,
      <p class='as-stacked-bar-widget__description as-body'>{this.description}</p>,
    ];
  }

  // private _renderGraph() {}

  private _renderLegend() {
    if (this.showLegend) {
      return <p>legend</p>
    }
  }
}

export interface StackedbarData {
  category: string;
  values: {
    [propName: string]: number;
  };
}
