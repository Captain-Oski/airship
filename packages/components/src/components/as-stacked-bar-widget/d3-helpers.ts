import { BaseType, Selection } from 'd3-selection';
import { ColumnData } from './as-stacked-bar-widget';

/**
 *
 * Create a `g` with the plot class where the columns will be drawn.
 */
export function createPlot(container: Selection<HTMLElement, {}, null, undefined>) {
  return container
    .append('g')
    .attr('class', 'plot')
    .selectAll('rect');
}

/**
 *
 * Create a `g` with the column class, this column has several rectangles inside.
 */
export function createColumn(
  element: Selection<HTMLElement, {}, null, undefined>): Selection<BaseType, {}, BaseType, {}> {
  return element
    .append('g')
    .attr('class', 'column')
    .selectAll('rect');
}

/**
 * Draw every rectangle inside a column.
 * This function only works when all the rectangles are the same type (positive/negative).
 */
export function drawColumn(
  element: Selection<BaseType, {}, BaseType, {}>,
  columnData: ColumnData[],
  yOffset: number,
  xOffset: number,
  colWidth: number) {

  element
    .data(columnData)
    .enter()
    .append('rect')
    .attr('x', xOffset)
    .attr('y', (d) => {
      const y = _computeY(d, yOffset);
      yOffset = d.negative ? yOffset + d.size : yOffset - d.size;
      return y;
    })
    .attr('width', colWidth)
    .attr('height', (d) => `${d.size}%`)
    .attr('fill', (d) => d.color);
}


function _computeY(data: ColumnData, origin: number): string {
  if (data.negative) {
    return `${origin}%`;
  }
  return `${(origin - data.size)}%`;
}


export default { createPlot, createColumn, drawColumn };
