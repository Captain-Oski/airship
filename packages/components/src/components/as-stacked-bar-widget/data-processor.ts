import { StackedbarData } from './as-stacked-bar-widget';

export function getDomain(data: StackedbarData[]): number[] {
  return data.reduce((previousValue: number[], currentValue: StackedbarData) => {
    return previousValue;
  }, [0, 0]);
}



export default { getDomain };
