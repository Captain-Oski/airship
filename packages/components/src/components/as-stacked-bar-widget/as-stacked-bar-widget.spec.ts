import { TestWindow } from '@stencil/core/dist/testing';
import { StackedBarWidget } from './as-stacked-bar-widget';

describe('as-stacked-bar-widget', () => {
  let stackedBarWidget;

  beforeEach(() => {
    stackedBarWidget = new StackedBarWidget();
  });

  it('should build', () => {
    expect(stackedBarWidget).toBeTruthy();
  });

  describe('Propery handling', () => {
    let testWindow;
    beforeEach(() => testWindow = new TestWindow());
    it('should handle title attribute', async () => {
      const element = await testWindow.load({
        components: [StackedBarWidget],
        html: `<as-stacked-bar-widget title="Population Income"></as-stacked-bar-widget>`
      });
      expect(element).toMatchSnapshot();
    });

    it('should handle description attribute', async () => {
      const element = await testWindow.load({
        components: [StackedBarWidget],
        html: `<as-stacked-bar-widget description="fake-description-attr"></as-stacked-bar-widget>`
      });
      expect(element).toMatchSnapshot();
    });

    it('should handle showLegend attribute', async () => {
      const element = await testWindow.load({
        components: [StackedBarWidget],
        html: `<as-stacked-bar-widget show-legend="false"></as-stacked-bar-widget>`
      });
      expect(element).toMatchSnapshot();
    });

    it('should handle data attribute', async () => {
      const element = await testWindow.load({
        components: [StackedBarWidget],
        html: `<as-stacked-bar-widget></as-stacked-bar-widget>`
      });
      element.data = 'fake_data';
      expect(element).toMatchSnapshot();
    });

    it('should handle valuesInfo attribute', async () => {
      const element = await testWindow.load({
        components: [StackedBarWidget],
        html: `<as-stacked-bar-widget></as-stacked-bar-widget>`
      });
      element.valuesInfo = 'fake_valuesInfo';
      expect(element).toMatchSnapshot();
    });
  });
});
