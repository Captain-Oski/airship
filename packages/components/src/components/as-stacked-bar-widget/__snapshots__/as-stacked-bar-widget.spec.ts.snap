// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-stacked-bar-widget Propery handling should handle data attribute 1`] = `
<as-stacked-bar-widget
  class="hydrated"
>
  
  <div
    class="as-widget"
  >
    <h2
      class="as-stacked-bar-widget__header"
    >
      
    </h2>
    <p
      class="as-stacked-bar-widget__description as-body"
    >
      
    </p>
    <svg />
    <as-y-axis
      from="0"
      to="0"
    />
    <p>
      legend
    </p>
  </div>
</as-stacked-bar-widget>
`;

exports[`as-stacked-bar-widget Propery handling should handle description attribute 1`] = `
<as-stacked-bar-widget
  class="hydrated"
  description="fake-description-attr"
>
  
  <div
    class="as-widget"
  >
    <h2
      class="as-stacked-bar-widget__header"
    >
      
    </h2>
    <p
      class="as-stacked-bar-widget__description as-body"
    >
      fake-description-attr
    </p>
    <svg />
    <as-y-axis
      from="0"
      to="0"
    />
    <p>
      legend
    </p>
  </div>
</as-stacked-bar-widget>
`;

exports[`as-stacked-bar-widget Propery handling should handle heading attribute 1`] = `
<as-stacked-bar-widget
  class="hydrated"
  heading="fake-heading-attr"
>
  
  <div
    class="as-widget"
  >
    <h2
      class="as-stacked-bar-widget__header"
    >
      fake-heading-attr
    </h2>
    <p
      class="as-stacked-bar-widget__description as-body"
    >
      
    </p>
    <svg />
    <as-y-axis
      from="0"
      to="0"
    />
    <p>
      legend
    </p>
  </div>
</as-stacked-bar-widget>
`;

exports[`as-stacked-bar-widget Propery handling should handle showLegend attribute 1`] = `
<as-stacked-bar-widget
  class="hydrated"
  show-legend="false"
>
  
  <div
    class="as-widget"
  >
    <h2
      class="as-stacked-bar-widget__header"
    >
      
    </h2>
    <p
      class="as-stacked-bar-widget__description as-body"
    >
      
    </p>
    <svg />
    <as-y-axis
      from="0"
      to="0"
    />
    
  </div>
</as-stacked-bar-widget>
`;

exports[`as-stacked-bar-widget Propery handling should handle valuesInfo attribute 1`] = `
<as-stacked-bar-widget
  class="hydrated"
>
  
  <div
    class="as-widget"
  >
    <h2
      class="as-stacked-bar-widget__header"
    >
      
    </h2>
    <p
      class="as-stacked-bar-widget__description as-body"
    >
      
    </p>
    <svg />
    <as-y-axis
      from="0"
      to="0"
    />
    <p>
      legend
    </p>
  </div>
</as-stacked-bar-widget>
`;
